### PR TITLE
Overhaul of Python API documentation

### PR DESCRIPTION
**Pull request by paulromano**
In anticipation of our upcoming PHYSOR workshop, I decided to make an attempt to improve our Python API documentation. This pull request makes a number of changes to that effect:

- Unrelated to the documentation itself -- the universe module is broken up into three modules: universe, cell, and lattice.
- Heavy use of sphinx.ext.autosummary and its autogen feature is made to create stubs for classes automatically. Now, we just list each class in an `.. autosummary::` table and a stub is created for it.
- I've also used sphinx.ext.intersphinx to create cross references to types from Python, numpy, and Pandas. For an example, see [Tally.get_pandas_dataframe()](http://web.mit.edu/romano7/www/html/pythonapi/generated/openmc.Tally.html#openmc.Tally.get_pandas_dataframe).
- Building docs with Python3 was broken and is now fixed.
- The Sphinx RTD theme is now used for local builds as well.
- Docstrings in the API were modified to use fully qualified paths for classes, e.g. `openmc.Filter` instead of just `Filter`.

You can see a preview of the docs [here](http://web.mit.edu/romano7/www/html/pythonapi/index.html).

@wbinventor Want to review?

### CODE REVIEW

**Comment by wbinventor** [comment]
I will take a look at this later today.

File: docs/source/pythonapi/index.rst [hunk 1]
```diff
@@ -13,61 +13,261 @@ online. We recommend going through the modules from Codecademy_ and/or the
...
+   :nosignatures:
+   :template: myclass.rst
+
+   openmc.XPlane
+   openmc.YPlane
+   openmc.ZPlane
+   openmc.XCylinder
+   openmc.YCylinder
+   openmc.ZCylinder
+   openmc.Sphere
```
Comment (wbinventor) [inline docs/source/pythonapi/index.rst]: Looks like we are missing `openmc.Cone` here? And it's subclasses `XCone`, `YCone` and `ZCone`.

File: openmc/__init__.py [hunk 2]
```diff
@@ -16,6 +18,10 @@
 from openmc.executor import *
 from openmc.statepoint import *
 from openmc.summary import *
+from openmc.region import *
+from openmc.source import *
+from openmc.particle_restart import *
+from openmc.arithmetic import *
```
Comment (wbinventor) [inline openmc/__init__.py]: Is `arithmetic` something we want imported for the user? I think this is more back-end code to keep hidden from the user.

File: openmc/__init__.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+from openmc.cell import *
```
Comment (wbinventor) [inline openmc/__init__.py]: Are these necessary? I thought they were imported with `openmc.geometry` below, but I may be wrong.

Comment (wbinventor) [inline openmc/__init__.py]: Oh nvm, I see that you've broken out `Cell` and `Lattice` into their own source files. This was a much needed change, I'm glad to finally see it in place!

File: openmc/cell.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+import warnings
+
+import openmc
+import openmc.checkvalue as cv
+from openmc.surface import Halfspace
+from openmc.region import Region, Intersection, Complement
+
+if sys.version_info[0] >= 3:
+    basestring = str
+
```
Comment (wbinventor) [inline openmc/cell.py]: Seems like there are a few extra newlines here and below `reset_auto_cell_id()`.

File: openmc/lattice.py [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+        return univs
+
+    def get_all_nuclides(self):
+        """Return all nuclides contained in the lattice
+
+        Returns
+        -------
+        nuclides : collections.OrderedDict
+            Dictionary whose keys are nuclide names and values are 2-tuples of
+            (nuclide, density)
```
Comment (wbinventor) [inline openmc/lattice.py]: should this be `(openmc.nuclide, density)`?

File: openmc/lattice.py [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+
+        return nuclides
+
+    def get_all_cells(self):
+        """Return all cells that are contained within the lattice
+
+        Returns
+        -------
+        cells : collections.OrderedDict
+            Dictionary whose keys are cell IDs and values are Cell instances
```
Comment (wbinventor) [inline openmc/lattice.py]: should this be "are openmc.Cell instances"?

File: openmc/lattice.py [hunk 7]
```diff
@@ -0,0 +1,869 @@
...
+
+        return cells
+
+    def get_all_materials(self):
+        """Return all materials that are contained within the lattice
+
+        Returns
+        -------
+        materials : collections.OrderedDict
+            Dictionary whose keys are material IDs and values are Material instances
```
Comment (wbinventor) [inline openmc/lattice.py]: ditto?

Comment (wbinventor) [inline openmc/lattice.py]: here and below

File: openmc/lattice.py [hunk 8]
```diff
@@ -0,0 +1,869 @@
...
+    name : str, optional
+        Name of the lattice. If not specified, the name is the empty string.
+
+    Attributes
+    ----------
+    id : int
+        Unique identifier for the lattice
+    name : str
+        Name of the lattice
+    dimension : array-like of int
```
Comment (wbinventor) [inline openmc/lattice.py]: what is "array-like"? why not just "Iterable" as we use elsewhere?

Comment (wbinventor) [inline openmc/lattice.py]: here and below

File: openmc/mgxs/library.py [hunk 9]
```diff
@@ -350,7 +350,7 @@ def build_library(self):
     def add_to_tallies_file(self, tallies_file, merge=True):
         """Add all tallies from all MGXS objects to a tallies file.
 
-        NOTE: This assumes that build_library() has been called
+        NOTE: This assumes that :meth:`Library.build_library` has been called
```
Comment (wbinventor) [inline openmc/mgxs/library.py]: typo: "meth" -> "math" (lol)

File: openmc/region.py [hunk 10]
```diff
@@ -255,21 +256,22 @@ class Union(Region):
...
-    >>> s2 = openmc.surface.Sphere(R=637.1e6)
+    >>> s1 = openmc.ZPlane(z0=0.0)
+    >>> s2 = openmc.Sphere(R=637.1e6)
     >>> type(-s2 | +s1)
     <class 'openmc.region.Union'>
 
     Parameters
     ----------
-    *nodes
+    \*nodes
```
Comment (wbinventor) [inline openmc/region.py]: is the backslash intentional?

File: openmc/mgxs/library.py (hunk 9 above)
Comment (paulromano) [inline openmc/mgxs/library.py]: It is actually [not a typo](http://www.sphinx-doc.org/en/stable/domains.html#role-py:meth).

Comment (wbinventor) [inline openmc/mgxs/library.py]: Oh! Well in that case, I see a number of places where you use `:math:` instead...perhaps you meant to use `:meth:`?

**Comment by wbinventor** [comment]
@paulromano all of this looks wonderful to me! I just built the updated docs on my machine and everything checks out. I only made a few very minor actionable comments - after they are addressed, I'll merge this in.

File: openmc/lattice.py (hunk 5 above)
Comment (paulromano) [inline openmc/lattice.py]: This one I'm going to leave as is -- if I'm referring to type I'd have to say `(openmc.Nuclide, float)` which seems weird.

File: openmc/lattice.py (hunk 6 above)
Comment (paulromano) [inline openmc/lattice.py]: The correct way to do this is ```are :class:`Cell` instances```.

File: openmc/lattice.py (hunk 8 above)
Comment (paulromano) [inline openmc/lattice.py]: Old usage -- I'll update it

File: openmc/mgxs/library.py (hunk 9 above)
Comment (paulromano) [inline openmc/mgxs/library.py]: Now *those* are typos!

File: openmc/region.py (hunk 10 above)
Comment (paulromano) [inline openmc/region.py]: Yes, otherwise Sphinx thinks I have an emphasis-open without an emphasis-close

**Comment by paulromano** [comment]
I believe all actionable comments have been addressed.

**Comment by wbinventor** [comment]
Thanks @paulromano!

**Pull request merged**